File: member4_map_matching/cpp/src/RoadpackValidate.cpp
#include "member4/roadpack_validate.hpp"

#include <cmath>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>

namespace sih26168::member4 {

RoadpackValidation validateRoadpackFile(const std::string& path, std::uint64_t max_bytes) {
    RoadpackValidation v;
    namespace fs = std::filesystem;
    std::error_code ec;
    if (!fs::exists(path, ec) || !fs::is_regular_file(path, ec)) {
        v.error = "roadpack missing or not a file";
        return v;
    }
    const auto sz = fs::file_size(path, ec);
    if (ec) {
        v.error = "unable to stat roadpack";
        return v;
    }
    v.bytes = static_cast<std::uint64_t>(sz);
    if (v.bytes == 0 || v.bytes > max_bytes) {
        v.error = "roadpack size rejected";
        return v;
    }

    std::ifstream in(path);
    if (!in) {
        v.error = "failed to open roadpack";
        return v;
    }
    std::string line;
    if (!std::getline(in, line) || line != "SIH26168_ROADPACK_V1") {
        v.error = "invalid roadpack header";
        return v;
    }
    v.schema = "SIH26168_ROADPACK_V1";

    bool have_bounds = false;
    auto touch = [&](double lat, double lon) {
        if (!std::isfinite(lat) || !std::isfinite(lon) || lat < -90.0 || lat > 90.0 ||
            lon < -180.0 || lon > 180.0) {
            return false;
        }
        if (!have_bounds) {
            v.min_lat = v.max_lat = lat;
            v.min_lon = v.max_lon = lon;
            have_bounds = true;
        } else {
            v.min_lat = std::min(v.min_lat, lat);
            v.max_lat = std::max(v.max_lat, lat);
            v.min_lon = std::min(v.min_lon, lon);
            v.max_lon = std::max(v.max_lon, lon);
        }
        return true;
    };

    while (std::getline(in, line)) {
        if (line.empty()) {
            continue;
        }
        std::istringstream ss(line);
        std::string tag;
        ss >> tag;
        if (tag == "NODE") {
            int id = 0;
            double lat = 0, lon = 0;
            ss >> id >> lat >> lon;
            if (!touch(lat, lon)) {
                v.error = "non-finite or out-of-range node";
                return v;
            }
            ++v.node_count;
        } else if (tag == "SEG") {
            ++v.edge_count;
        } else if (tag == "PT") {
            double lat = 0, lon = 0;
            ss >> lat >> lon;
            if (!touch(lat, lon)) {
                v.error = "non-finite or out-of-range geometry";
                return v;
            }
        }
    }
    if (v.edge_count <= 0 || v.node_count <= 0 || !have_bounds) {
        v.error = "roadpack failed graph sanity (nodes/edges/bounds)";
        return v;
    }
    v.ok = true;
    return v;
}

}  // namespace sih26168::member4

File: member4_map_matching/python/sih26168_map_matching/roadpack_validate.py
"""Validate untrusted .roadpack files. No execution, no network."""

from __future__ import annotations

import hashlib
import math
from dataclasses import dataclass
from pathlib import Path

from .road_graph import load_roadpack


@dataclass
class ValidationResult:
    ok: bool
    error: str = ""
    node_count: int = 0
    edge_count: int = 0
    bytes: int = 0
    checksum_sha256: str = ""
    min_lat: float = 0.0
    max_lat: float = 0.0
    min_lon: float = 0.0
    max_lon: float = 0.0


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()


def validate_roadpack(
    path: Path,
    *,
    expected_checksum: str | None = None,
    max_bytes: int = 64 * 1024 * 1024,
) -> ValidationResult:
    if not path.is_file():
        return ValidationResult(False, "roadpack missing")
    size = path.stat().st_size
    if size <= 0 or size > max_bytes:
        return ValidationResult(False, "roadpack size rejected", bytes=size)
    header = path.read_text(encoding="utf-8", errors="replace").splitlines()[:1]
    if not header or header[0].strip() != "SIH26168_ROADPACK_V1":
        return ValidationResult(False, "invalid roadpack header", bytes=size)
    digest = sha256_file(path)
    if expected_checksum and digest.lower() != expected_checksum.lower():
        return ValidationResult(False, "checksum mismatch", bytes=size, checksum_sha256=digest)
    try:
        network = load_roadpack(path)
    except Exception as exc:  # noqa: BLE001 — treat any parse error as invalid input
        return ValidationResult(False, f"schema/parse error: {exc}", bytes=size, checksum_sha256=digest)
    if not network.segments or not network.node_coords:
        return ValidationResult(False, "empty graph", bytes=size, checksum_sha256=digest)
    bounds = network.bounds()
    if bounds is None:
        return ValidationResult(False, "missing bounds", bytes=size, checksum_sha256=digest)
    min_lat, max_lat, min_lon, max_lon = bounds
    for lat, lon in network.node_coords.values():
        if not math.isfinite(lat) or not math.isfinite(lon):
            return ValidationResult(False, "NaN/Inf coordinate", bytes=size, checksum_sha256=digest)
        if not -90.0 <= lat <= 90.0 or not -180.0 <= lon <= 180.0:
            return ValidationResult(False, "coordinate out of range", bytes=size, checksum_sha256=digest)
    connected = network.graph.number_of_edges()
    if connected <= 0:
        return ValidationResult(False, "no connectivity", bytes=size, checksum_sha256=digest)
    return ValidationResult(
        True,
        node_count=len(network.node_coords),
        edge_count=len(network.segments),
        bytes=size,
        checksum_sha256=digest,
        min_lat=min_lat,
        max_lat=max_lat,
        min_lon=min_lon,
        max_lon=max_lon,
    )
